### PR TITLE
Broadcast support

### LINKED ISSUE #511: Tests for Broadcast
Now that we are [adding broadcast support in the catalog](https://github.com/uwescience/myria/issues/814), I suspect that many tests would break if they were performed on broadcasted input relations. The reason is that Raco doesn't reason in a robust way about broadcasted relations.

### PR DESCRIPTION
**Pull request by bmyerz**
fixes #511

### CODE REVIEW

**Comment by bmyerz** [comment]
current state: The functionality should work in many cases. The broadcast plans are not yet thoroughly tested. I'm particularly concerned with the set operators. 

The best tests will be those that show executing the plan gives the correct results. This requires either tests with Myria or enhancing the FakeDB evaluator to consider physical properties.

**Comment by senderista** [comment]
OK, should I pick it up from here (i.e., add tests in MyriaX and/or raco and fix any resulting issues)?

**Comment by bmyerz** [comment]
If we use MyriaX for tests of broadcast, they will have to be system tests, since it involves both repositories. [These examples](https://github.com/uwescience/raco/blob/master/docs/index.md#use-raco-from-python) provide a skeleton of what might go in the test code (using Myria through Raco).

**Comment by bmyerz** [comment]
The failing test on [the last commit](https://github.com/uwescience/raco/pull/517/commits/01a692c869fbc7567b9940e6e306adc990454176)

image

indicates that my DeBroadcast operator is interfering with push Into SQL. It seems like PushIntoSQL rule should go first, following by PushDeBroadcast. PushDeBroadcast will then potentially be pushed into MyriaQueryScan, which enables the worker-0-only optimization.

File: raco/rules.py [hunk 1]
```diff
@@ -1041,3 +1043,55 @@ def check_partition_equality(op, representation):
...
+                copy = MoveDeBroadcast._copy_op(expr.input)
+                copy._debroadcast = True
+                return copy
+            elif isinstance(expr.input, algebra.UnaryOperator):
+                copy = MoveDeBroadcast._copy_op(expr.input)
+                exprcopy = MoveDeBroadcast._copy_op(expr)
+                exprcopy.input = copy.input
+                copy.input = exprcopy
+                return copy
+            elif isinstance(expr.input, algebra.BinaryOperator):
```
Comment (senderista) [inline raco/rules.py]: add comment that this case pushes a DeBroadcast op down one level on each branch

File: raco/backends/myria/myria.py [hunk 2]
```diff
@@ -1663,6 +1706,7 @@ def fire(self, expr):
     rules.OneToOne(algebra.Difference, MyriaDifference),
     rules.OneToOne(algebra.OrderBy, MyriaInMemoryOrderBy),
     rules.OneToOne(algebra.Sink, MyriaSink),
+    DeBroadcastToSelect()
```
Comment (senderista) [inline raco/backends/myria/myria.py]: Do we actually need this rule given the `MoveDeBroadcast` rule, which pushes `DeBroadcast` ops all the way down to scans, and then sets `overrideWorkers` on those scans to `[0]`? The `DeBroadcastToSelect` rule seems very hackish by comparison.

**Comment by senderista** [comment]
I've addressed my own comments and added enhancements on the `myria_broadcast` branch (which is compatible with the Myria `add_broadcast` branch).

**Comment by senderista** [comment]
superseded by https://github.com/uwescience/raco/pull/548

**Pull request closed without merging**